### PR TITLE
[`flake8-pyi`] Implement PYI055

### PR DESCRIPTION
**Pull request by LaBatata101**
## Summary

Checks for the presence of multiple `type`s in a union. See [original source](https://github.com/PyCQA/flake8-pyi/blob/2a86db8271dc500247a8a69419536240334731cf/pyi.py#L1315).

ref #848 

## Test Plan

Snapshots and manual runs flake8.

### CODE REVIEW

**Comment by zanieb** [comment]
Let's turn this on for `.py` files too now that we've done https://github.com/astral-sh/ruff/pull/6297

**Comment by LaBatata101** [comment]
> Let's turn this on for `.py` files too now that we've done #6297

Working on it!

File: crates/ruff/src/checkers/ast/analyze/expression.rs [hunk 1]
```diff
@@ -96,8 +97,13 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
                     if checker.enabled(Rule::DuplicateUnionMember) {
                         flake8_pyi::rules::duplicate_union_member(checker, expr);
                     }
-                    if checker.is_stub && checker.enabled(Rule::RedundantLiteralUnion) {
-                        flake8_pyi::rules::redundant_literal_union(checker, expr);
+                    if checker.is_stub {
+                        if checker.enabled(Rule::RedundantLiteralUnion) {
+                            flake8_pyi::rules::redundant_literal_union(checker, expr);
+                        }
```
Comment (LaBatata101) [inline crates/ruff/src/checkers/ast/analyze/expression.rs:103]: This was forgotten in #6297 I will be fixing in this PR if that's ok

File: crates/ruff/src/checkers/ast/analyze/expression.rs [hunk 2]
```diff
@@ -1103,6 +1105,15 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
             {
                 flake8_pyi::rules::redundant_literal_union(checker, expr);
             }
+            if checker.enabled(Rule::UnnecessaryTypeUnion)
+              // Avoid duplicate checks if the parent is an `|`
+                && !matches!(
+                    checker.semantic.expr_parent(),
+                    Some(Expr::BinOp(ast::ExprBinOp { op: Operator::BitOr, ..}))
+                )
```
Comment (zanieb) [inline crates/ruff/src/checkers/ast/analyze/expression.rs:1113]: Should we consider moving this check into a shared block like we do for the `Union[...]` case above?

File: crates/ruff/src/rules/flake8_pyi/rules/unnecessary_type_union.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        let Some(call_path) = checker.semantic().resolve_call_path(subscript.value.as_ref()) else {
+            return;
+        };
+
+        if matches!(call_path.as_slice(), ["" | "builtins", "type"]) {
+            type_exprs.push(&subscript.slice);
+        }
+    };
+
+    traverse_union(&mut collect_type_exprs, checker.semantic(), union, None);
```
Comment (zanieb) [inline crates/ruff/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:68]: How was using this helper? I wrote it and would appreciate any feedback :)

File: crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI051_PYI051.py.snap [hunk 4]
```diff
@@ -11,6 +11,62 @@ PYI051.py:4:18: PYI051 `Literal["foo"]` is redundant in a union with `str`
 6 | C: TypeAlias = typing.Union[Literal[5], int, typing.Union[Literal["foo"], str]]
   |
 
+PYI051.py:5:37: PYI051 `Literal[b"bar"]` is redundant in a union with `bytes`
```
Comment (zanieb) [inline crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI051_PYI051.py.snap:14]: Oops guessing this snapshot isn't meant to be committed here

Comment (zanieb) [reply crates/ruff/src/rules/flake8_pyi/snapshots/ruff__rules__flake8_pyi__tests__PYI051_PYI051.py.snap:14]: (Nevermind I see this is for the py file)

**Review by zanieb — APPROVED** [review]
LGTM! Thank you :)

File: crates/ruff/src/checkers/ast/analyze/expression.rs (hunk 2 above)
Comment (LaBatata101) [reply crates/ruff/src/checkers/ast/analyze/expression.rs:1113]: That would be nice, removing the duplicated code. Does the variable name `is_unchecked_union` fit here as well?

File: crates/ruff/src/rules/flake8_pyi/rules/unnecessary_type_union.rs (hunk 3 above)
Comment (LaBatata101) [reply crates/ruff/src/rules/flake8_pyi/rules/unnecessary_type_union.rs:68]: It was very easy to use, good work!

**Pull request merged**